### ISSUE TITLE
Support mixed precision calculation

### ISSUE THREAD

**Issue by mathpluscode** [DeepRegNet/DeepReg#468]
## Subject of the feature

Using mixed precision (float16 and float32) can save GPU memory and hopefully the calculation speed as well.

According to https://www.tensorflow.org/guide/mixed_precision, the modification is not too complicated.

Some basic facts:

1. Computations are done in float16 for performance, but variables must be kept in float32 for numeric stability. They cast their inputs to float16 in order to do float16 computations, which causes their outputs to be float16 as a result.
2. Even if the model does not end in a softmax, the outputs should still be float32.
3. In general you should run input processing math in float32 if it runs on the CPU.
4. Custom loss should be calculated in float32

**Comment by mathpluscode** [comment]
Now I can confirm that we will need multiple changes to support mixed precision calculation. It is not as simple as switching the policy.
